### ISSUE TITLE
Add ability to automatically globally catch all database errors

### ISSUE THREAD

**Issue by D-Marc1** [adonisjs/core#3372]
## Why this feature is required (specific use-cases will be appreciated)?

In production, if there's a database error, it'll at least hide the stack trace, but still show the message with all database fields affected and other private database info in the response (this should never happen). Obviously you can meticulously `try/catch` every database action you do, but this would be nice for it to not be required in cases where a generic error message to the user is good enough.

It would be nice if there could either be a unified `DatabaseException` type or if it at least could indicate in some way that this is specifically a database error, so you wouldn't have to meticulously `try/catch` every single database action. Even in situations where you need to give a specific message, you can easily just `try/catch` it, re-throw the error and add any additional message to pass to the global exception handler.

Alternatively, I wonder if `error.message` even needs to list the detailed, private database error, when this is already in the stack trace anyway, which is hidden by default in production. This would be another way of dealing with the issue of it automatically leaking database info in production.

## Have you tried any other work arounds?

Right now, I do something like the following for every database action:

```ts
try {
  await User.create({})
} catch (error) {
  throw Object.assign(error, { type: 'DatabaseError' })
}
```

Then I catch it in the global exception handler with the following;

```ts
public async handle(error, ctx) {
  if (error.type === 'DatabaseError') {
    // PostgreSQL Unique Constraint Violation
    if (error.code === '23505') {
      return ctx.response.status(409)
    } else if (error.code === 'E_ROW_NOT_FOUND') {
      return ctx.response.status(404).send('Row not found')
    }
    
    // Default database error response
    return ctx.response.status(500)
  }

  return super.handle(error, ctx)
}
```

## Are you willing to work on it with little guidance?

I could potentially be willing to if I have the time to do it.

**Comment by thetutlage** [comment]
Yeah, not a bad idea to unify all errors inside a single DatabaseException class. 

But, for the starters you can simply show a generic message for all errors which doesn't have a status or status is >500. This way even beyond database you will not leak any server related exceptions to the user.

Back to the database exception. It will be nice if we can find some Knex API that allows catching and re-throwing exceptions globally
